### ISSUE TITLE
Global Timeout using Junit Rule

### ISSUE THREAD

**Issue by kishoretak** [Nordstrom/JUnit-Foundation#57]
Hi,
Currently, this library is overriding _timeout_ annotation to have a global timeout.
Junit has deprecated this option and introduced Timeout Rule, which also considers the time spent in the _before_ and _after_ method along with the unit test's time. (_timeout_ annotation doesn't consider _before_ and _after_ method time).
I've made this change in the library via overriding _getTestRules_ method of BlockJUnit4ClassRunner and adding _timeout_ rule to existing rule list.
Please share your thought on same.

Thanks

**Comment by sbabcoc** [comment]
I considered using the **Timeout** rule to implement the global timeout feature, but this mechanism doesn't provide the necessary level of control. The primary issue is that the **Timeout** rule sets the maximum duration of the entire atomic test, including @Before and @<~-- -->After methods. That behavior means that tests can time out because of delays in setup of teardown, which aren't actually part of the test.

 If you look [here](https://github.com/junit-team/junit4/wiki/Timeout-for-tests), you'll see that I added information to this JUnit 4 Wiki page _after_ the information about the **Timeout** rule.
As far as I know, the _timeout_ element of the @Test annotation of JUnit 4 has never been deprecated.

**Comment by kishoretak** [comment]
Hi @sbabcoc 
Thanks for your prompt response!!
- The primary issue is that the Timeout rule sets the maximum duration of the entire atomic test, 
    including @Before and @After methods.

This is exactly what we expect as we want to enforce the overall timeout, I agree that it might not be correct behavior for your use-case. I wanted to know your thoughts on my approach for our use-case.

- As far as I know, the timeout element of the @Test annotation of JUnit 4 has never been deprecated.

Yes, the timeout element of the @Test annotation of JUnit 4 is not deprecated. I was referring to this - https://github.com/junit-team/junit4/blob/6d0fad48ce3a05b32d903d2016c24d276b6e1eb8/src/main/java/org/junit/runners/BlockJUnit4ClassRunner.java#L356

**Comment by sbabcoc** [comment]
Weird that they deprecated a protected method that almost no one will even notice, but failed to deprecate the _timeout_ element of the @Test annotation that relates to it.
It's easy to apply a global timeout through a proxy of the @Test annotation, but auto-applied test rules are another matter entirely. There's no way to exempt specific classes and methods from the effects of the Timeout rule. I'd also need to dig into the guts of JUnit 4 to figure out how a framework extension like **JUnit Foundation** would go about adding a test rules automatically.

**Comment by kishoretak** [comment]
I'm overriding the _getTestRules_ method of _BlockJUnit4ClassRunner_ and appending a new timeout rule to existing rules.
https://github.com/junit-team/junit4/blob/master/src/main/java/org/junit/runners/BlockJUnit4ClassRunner.java#L434

`package com.nordstrom.automation.junit;

import net.bytebuddy.implementation.bind.annotation.Argument;
import net.bytebuddy.implementation.bind.annotation.RuntimeType;
import net.bytebuddy.implementation.bind.annotation.SuperCall;
import net.bytebuddy.implementation.bind.annotation.This;
import org.junit.rules.TestRule;
import org.junit.rules.Timeout;

import java.util.List;
import java.util.Objects;
import java.util.concurrent.Callable;

public class GetTestRules {
    /**
     * Interceptor for the {@link org.junit.runners.BlockJUnit4ClassRunner#getTestRules(Object)}  getTestRules} method.
     *
     * @param runner target {@link org.junit.runners.BlockJUnit4ClassRunner BlockJUnit4ClassRunner} object
     * @param proxy  callable proxy for the intercepted method
     * @param target the test case instance
     * @return {@code anything} - JUnit test class instance
     * @throws Exception {@code anything} (exception thrown by the intercepted method)
     */
    @RuntimeType
    public static List<TestRule> intercept(@This final Object runner, @SuperCall final Callable<?> proxy, @Argument(0) final Object target) throws Exception {
        List<TestRule> testRules = (List<TestRule>) LifecycleHooks.callProxy(proxy);
        Long timeout = null;
        try {
            timeout = Long.getLong("junitTestTimeoutInMilliSeconds");
            System.out.println("junitTestTimeoutInMilliSeconds = " + timeout);
        } catch (Exception ignored) {
        }
        if (!Objects.isNull(timeout)) {
            testRules.add(Timeout.millis(timeout));
        }
        return testRules;
    }
}
`

**Comment by sbabcoc** [comment]
Good work! This provides a great starting point. Prior to adding the test rule, you should provide the following behavior:
# Does the test object already have a **Timeout** rule attached to it?
  # YES => If the existing **Timeout** rule specifies a longer interval than the default test timeout, leave it alone.
  # NO => Replace the existing **Timeout** rule with a new one built with the default timeout.
# Does the @Test annotation of the method for which the target test object was created have the _timeout_ element?
  # YES => Proxy the @Test annotation with one that removes the _timeout_ element.

**Comment by kishoretak** [comment]
Thanks @sbabcoc
Regarding point 2.
Does the @Test annotation of the method for which the target test object was created have the 
   timeout element?
YES => Proxy the @Test annotation with one that removes the timeout element.

You mean if a test already has timeout arg set in @Test annotation then remove it?

Few more point to consider -
1. We'll be using both timeout annotation and timeout rule for global timeout?
2. YES -  Do we want to use the same global timeout parameter (junit.timeout.test) to be used in both timeout annotation and timeout rule or two separate config params?
3. NO - we'll remove timeout annotation and use only timeout rule going forward for the global timeout.

This will derive the overall strategy for the final (impacting) timeout which a test going to consider.

**Comment by sbabcoc** [comment]
Both timeout features can be active concurrently. The _timeout_ element of the @Test annotation limits the duration of the test method in isolation; the **Timeout** rule limits the duration of the atomic test, which includes all setup and cleanup methods. It's very easy to specify conflicting values if management is provided for both timeout features, but I guess this falls under the heading of `caveat emptor`.
The current management feature enables global control of test method timeout via the **TEST_TIMEOUT** configuration option. I propose adding a new **TIMEOUT_RULE** option to provide management of your **Timeout** rule feature.
When applying the **TEST_TIMEOUT** option:
- Does the @Test annotation specify the _timeout_ element?
  - YES => If the existing _timeout_ element specifies a longer interval (**or _0_ for no timeout**), leave it alone.
  - NO (or YES with shorter non-zero interval) => Proxy the @Test annotation, assigning the global _timeout_ interval.
When applying the **TIMEOUT_RULE** option:
- Does the test object already have a **Timeout** rule attached to it?
  - YES => If the existing **Timeout** rule specifies a longer interval than the default test timeout, leave it alone.
  - NO => Replace the existing **Timeout** rule with a new one built with the default timeout.

**Comment by sbabcoc** [comment]
I've added management of a global Timeout rule. I'll be adding the ability to disable timeouts entirely.

**Comment by kishoretak** [comment]
Thanks @sbabcoc, actually I wanted to merge this change from my end as part of my contribution to open-source :)

**Comment by sbabcoc** [comment]
Sorry, @kishoretak! Your feature suggestion was excellent, and the code you contributed was a great starting point. I ended up creating my own implementation of this feature, but I should have handled this set of revisions so that you got the contributor credit you deserve. I just released a new version that consolidates the two flavors of JUnit timeout control much better. The README documentation of timeout management could probably use some attention, and a pull request for this content would give you contributor credit.

**Comment by kishoretak** [comment]
Thanks, @sbabcoc!
I've identified one use-case which is kind of a feature request, currently, the library doesn't support timeout if the test is using [PowerMockRunner](https://www.javadoc.io/doc/org.powermock/powermock-module-junit4/1.6.3/org/powermock/modules/junit4/PowerMockRunner.html) (which is pretty common). I've added this functionality to my repo and want to check whether it's a correct use-case for this library. 
Let me know your thoughts on the same.

**Comment by sbabcoc** [comment]
@kishoretak 
As far as I know, alternate test runners like **PowerMockRunner** should be supported. This is potentially an issue is the way I installed the interceptors, though. Can you share a sample project that reproduces the issue?

**Comment by kishoretak** [comment]
@sbabcoc 
It should be reproducible for any test class defining @RunWith(PowerMockRunner.class).
As the Powermock runner doesn't extend the ParentRunner, instead it extends Runner abstract class and we'are overriding the behavior of ParentRunner and it's subclasses.

Also, there is an issue with using timeout parameter in @Test Annotation with PowerMockRunner - https://github.com/powermock/powermock/issues/817

**Comment by sbabcoc** [comment]
Any reference to **PowerMockRunner** is sufficient to disable the intercepts being installed by **JUnit Foundation**, because the PowerMock library is loading JUnit core classes before we can add our hooks. I don't what we could do to fix this.

**Comment by sbabcoc** [comment]
It's possible that I could create a wrapper for custom runners that would postpone loading the Unit classes until after we installed our hooks. I'll throw something together to see if this is a successful technique.

**Comment by sbabcoc** [comment]
Here's an example of a runner that existed in **JUnit Foundation** a long time ago: [HookInstallingRunner.java](https://github.com/Nordstrom/JUnit-Foundation/blob/ad1f1e88638edeff23093e391c10b23bb2e06eef/src/main/java/com/nordstrom/automation/junit/HookInstallingRunner.java)

**Comment by kishoretak** [comment]
Yeah, my approach is similar to above.

```
 .type(nameContains("PowerMockJUnit47MethodRunner"))
                .transform((builder, type, classloader, module) -> builder
                       .method(named("createStatement")).intercept(MethodDelegation.to(createStatement))
                        .implement(Hooked.class))
```
But as the junit-foundation is only meant to be helper for junit and not for powermock, So I'm not sure whether it should hanlde powermock cases or not.

**Comment by sbabcoc** [comment]
Yes, **JUnit Foundation** is a general-purpose solution. PowerMock is using deprecated JUnit interfaces, which JUnit Foundation doesn't currently support. This is the real reason the installed hooks are never invoked. The wrapper approach I was exploring is insufficient to add support for PowerMock.

The interfaces being used by PowerMock have been deprecated since JUnit 4.4. The **ClassRoadie** class, which PowerMock uses in its runner implementations, includes this comment:

> Included for backwards compatibility with JUnit 4.4. Will be  removed in the next major release. Please use _BlockJUnit4ClassRunner_ in place of_JUnit4ClassRunner_.

JUnit Foundation extends the current JUnit framework, and adding support for the deprecated interfaces would require a lot of work. Rules weren't added until JUnit 4.7, which is why PowerMock support for them is so uneven (they just hacked in as best they could).

You may want to explore switching over to [JMockit](http://jmockit.github.io) instead. This library is under active development, with support for JUnit 4, JUnit 5, and TestNG.

**Comment by kishoretak** [comment]
@sbabcoc, I totally agree with you on PowerMock using deprecated JUnit interfaces (ClassRoadie and MethodRoadie).
I'll explore JMockit for our use-case, thanks for your suggestion.
